File: 05-behavioral-patterns/04-mediator/04-mediator.go
package main

import "fmt"

type ChatRoomMediator interface {
	ShowMessage(sender *User, message string)
	Register(user *User)
}

type ChatRoom struct {
	users []*User
}

func (c *ChatRoom) Register(user *User) {
	c.users = append(c.users, user)
}

func (c *ChatRoom) ShowMessage(sender *User, message string) {
	for _, user := range c.users {
		if user != sender {
			user.Receive(sender.name, message)
		}
	}
}

type User struct {
	name     string
	chatroom ChatRoomMediator
}

func NewUser(name string, chatroom ChatRoomMediator) *User {
	user := &User{name: name, chatroom: chatroom}
	chatroom.Register(user)
	return user
}

func (u *User) Send(message string) {
	fmt.Printf("%s sends: %s\n", u.name, message)
	u.chatroom.ShowMessage(u, message)
}

func (u *User) Receive(senderName, message string) {
	fmt.Printf("%s receives from %s: %s\n", u.name, senderName, message)
}

func main() {
	chatroom := &ChatRoom{}
	alice := NewUser("Alice", chatroom)
	bob := NewUser("Bob", chatroom)
	carol := NewUser("Carol", chatroom)

	alice.Send("Hi everyone!")
	bob.Send("Hey Alice!")
}

File: 05-behavioral-patterns/04-mediator/04-mediator.py
class ChatRoomMediator:
    def show_message(self, sender, message):
        pass

class ChatRoom(ChatRoomMediator):
    def __init__(self):
        self.users = []

    def register(self, user):
        self.users.append(user)

    def show_message(self, sender, message):
        for user in self.users:
            if user != sender:
                user.receive(sender.name, message)

class User:
    def __init__(self, name, chatroom):
        self.name = name
        self.chatroom = chatroom
        self.chatroom.register(self)

    def send(self, message):
        print(f"{self.name} sends: {message}")
        self.chatroom.show_message(self, message)

    def receive(self, sender_name, message):
        print(f"{self.name} receives from {sender_name}: {message}")

# Client code
chatroom = ChatRoom()
alice = User("Alice", chatroom)
bob = User("Bob", chatroom)
carol = User("Carol", chatroom)

alice.send("Hi everyone!")
bob.send("Hello Alice!")
